latch_cli/services/cp/upload.py

```diff
@@ -0,0 +1,366 @@
+):
+    src_path = Path(src)
+    if not src_path.exists():
+        raise ValueError(f"Could not find {src_path}.")
+
+    normalized = normalize_path(dest)
+    dest_data = get_node_data(normalized, allow_resolve_to_parent=True)
+
+    dest_exists = not dest_data.is_parent
+    dest_is_dir = dest_data.type in {
+        LDataNodeType.account_root,
+        LDataNodeType.mount,
+        LDataNodeType.dir,
+    }
+
+    if config.progress == Progress.none:
+        num_bars = 0
+        show_total_progress = False
+    elif not src_path.is_dir():
+        num_bars = 1
+        show_total_progress = False
+    elif config.progress == Progress.total:
+        num_bars = 0
+        show_total_progress = True
+    else:
+        num_bars = config.max_concurrent_files
+        show_total_progress = True
+
+    with (
+        ProcessPoolExecutor() as executor,
+        ProgressBarManager() as manager,
+    ):
+        if src_path.is_dir():
+            if not dest_is_dir:
+                raise ValueError(f"{normalized} is not a directory.")
+
+            if dest_exists and not src.endswith("/"):
+                normalized = remote_joinpath(normalized, src_path.name)
+
+            srcs: List[Path] = []
+            dests: List[str] = []
+            total_bytes = 0
+
+            for dir_path, _, file_names in os.walk(src_path, followlinks=True):
+                for file_name in file_names:
+                    rel_path = Path(dir_path) / file_name
+
+                    srcs.append(rel_path)
+                    dests.append(
+                        remote_joinpath(normalized, str(rel_path.relative_to(src_path)))
+                    )
+                    total_bytes += rel_path.stat().st_size
+
+            num_files = len(srcs)
+
+            url_generation_bar: ProgressBars = manager.ProgressBars(
+                0, show_total_progress=(config.progress != Progress.none)
+            )
+            url_generation_bar.set_total(num_files, "Generating URLs")
+
+            start_upload_futs: List[Future[Optional[StartUploadReturnType]]] = []
+
+            start = time.perf_counter()
+            for s, d in zip(srcs, dests):
+                start_upload_futs.append(
+                    executor.submit(
+                        start_upload, s, d, config.chunk_size, url_generation_bar
+                    )
+                )
+
+            wait(start_upload_futs)
+            url_generation_bar.close()
+
+            chunk_upload_bars: ProgressBars = manager.ProgressBars(
+                min(num_bars, num_files),
+                show_total_progress=show_total_progress,
+                verbose=config.verbose,
+            )
+            chunk_upload_bars.set_total(num_files)
+
+            # todo(ayush): this is jank and also gives the illusion of synchronous execution
+            # since nothing gets interleaved
+            #
+            # perhaps do each file in its own process and have the chunks be uploaded with asyncio?
```

Comment:
I would do the whole thing in asyncio including the start upload.

Reply:
Also what do you mean by `gives the illusion of synchronous execution`? Do you mean `gives the illusion of asynchronous execution`? If so, why is it fully synchronous? this is using a process pool, right?

Reply:
https://pypi.org/project/requests-async/ seems to have a reasonable api. could be emulate with our tinyrequests?

Reply:
empirically it seems like the executor schedule jobs in the order in which they are submitted - every files' chunk jobs are submitted at once, so it will do one file at a time, chunks in parallel - hence it looks like execution is serial since only one file is being uploaded at a time (unless the number of chunks is < `max_workers`)

Reply:
it's completely correct behavior tho. otherwise ud spend hours uploading the first part of every file with none of them actually finishing

the tasks pbars should just be acquired per part and not per file